Introduce Floating Links in the Business Directory Panel

shervElmi opened the pull request:
This PR introduces the Floating Links feature to the Business Directory Panel, which provides users with a set of links for quick navigation to the **Upgrade**, **Support**, and **Documentation** pages.

#### NOTE: To test the `PRO version links`, activate the BD PRO plugin.

## Testing Instructions:
1. Navigate to `WP Admin > Directory`.
2. Check the perfect pixel implementation.
  - **Figma File:** https://www.figma.com/file/YcSWkrjVGroC1zyxhuHH6W/Floating-Links?node-id=1-117&t=Xf5EKO8Mvs1JwuFj-4
3. Verify that all functionality works well.

## Demo - BD Free Version:
![2023-07-03 16 16 19](https://github.com/Strategy11/business-directory-plugin/assets/69119241/43d6f610-dd14-4d0a-b662-1361b77add3a)

## Demo - BD PRO Version:
![2023-07-03 16 29 20](https://github.com/Strategy11/business-directory-plugin/assets/69119241/df836213-4759-4638-9244-e651a02daa79)

garretlaxton commented:
With the pro version the notification count still shows even when it's 0.
![image](https://github.com/Strategy11/business-directory-plugin/assets/16290508/033e8c65-e240-4759-837b-3ae0c6e62eda)
![image](https://github.com/Strategy11/business-directory-plugin/assets/16290508/23d26467-1272-476c-982e-0e650f63b3ad)

shervElmi commented:
> With the pro version the notification count still shows even when it's 0. ![image](https://user-images.githubusercontent.com/16290508/250649966-033e8c65-e240-4759-837b-3ae0c6e62eda.png) ![image](https://user-images.githubusercontent.com/16290508/250649996-23d26467-1272-476c-982e-0e650f63b3ad.png)


Thank you for your review, @garretlaxton. I haven't been able to reproduce this issue. Could you kindly provide a QA URL where this issue is apparent?

garretlaxton commented:
@shervElmi Sure!

https://bd.grunge.us.eu.org/wp-admin/
User: Sherv
Pass: vb3AwO#v5j#RNJN6fFBPU#I!

shervElmi commented:
> @shervElmi Sure!
> 
> https://bd.grunge.us.eu.org/wp-admin/ User: Sherv Pass: vb3AwO#v5j#RNJN6fFBPU#I!

Thanks a lot, @garretlaxton! I've fixed the issue you mentioned, and now the notifications box shows up when the page first loads.

**Note:** Please update your QA BD plugin.

garretlaxton commented:
Thanks! I see the notifications open now though.
![image](https://github.com/Strategy11/business-directory-plugin/assets/16290508/f3d4e9c3-aec1-462f-b836-8f39bed004aa)
![image](https://github.com/Strategy11/business-directory-plugin/assets/16290508/8e563d23-4df9-49bc-b6e3-2f8d8679b98e)

shervElmi commented:
> Thanks! I see the notifications open now though. ![image](https://user-images.githubusercontent.com/16290508/251731131-f3d4e9c3-aec1-462f-b836-8f39bed004aa.png) ![image](https://user-images.githubusercontent.com/16290508/251731266-8e563d23-4df9-49bc-b6e3-2f8d8679b98e.png)

Thanks! I've checked your QA site, and it seems that the latest updates haven't been applied. I tried clearing the cache, but some code still doesn't appear to be updated.

garretlaxton commented:
I primarily use my localhost for testing, I just use that site for sharing what I see. I have since updated the plugin there if you want to take a look now.

shervElmi commented:
Thanks, @garretlaxton for your review. I've just updated the code to stop showing the `Hide Notifications` button when there's nothing to display.

garretlaxton reviewed (approved):
Nice, this looks good now!

mostafa-hisham commented on includes/class-assets.php:
```diff
@@ -485,4 +488,35 @@ private static function is_jquery_ui_css_loaded() {
+	 * @param string $version Current version of the plugin.
+	 * @return void
+	 */
+	private static function enqueue_floating_links( $plugin_url, $version ) {
```
Thanks for your work @shervElmi 

I don't think we need to make this function static

shervElmi replied:
Thank you for your review, @mostafa-hisham!

We apply Floating Links in all S11 plugins. I made this method static to avoid using any `$this` reference in the future. I believe the static approach works well here.

stephywells commented on assets/css/packages/s11-floating-links.css:
```diff
@@ -0,0 +1,197 @@
+	display: flex;
+}
+
+.s11-floating-links-logo-icon {
```
The X and Percie aren't quite centered vertically in the circle. This css fixes it for me:
`.s11-svg-wrapper{display:flex}`

stephywells commented on assets/css/packages/s11-floating-links.css:
```diff
@@ -0,0 +1,197 @@
+
+.s11-floating-links {
+	position: fixed;
+	right: 48px;
```
Can we override this position in BD? I think it's 48px in FF because of the screen options tab. But here, it would looks better if it aligned with the notifications. It matches for me with 20px for right and bottom.

Before:
<img width="406" alt="Screenshot 2023-08-04 at 6 41 23 PM" src="https://github.com/Strategy11/business-directory-plugin/assets/1116876/9af83499-48ef-468c-8e6c-a9335ebe6dbe">

After:
<img width="358" alt="Screenshot 2023-08-04 at 6 41 11 PM" src="https://github.com/Strategy11/business-directory-plugin/assets/1116876/7ed5e4b5-dae2-42fc-b98d-ee4165708450">

stephywells replied:
This is almost there. It's just a bit too high now. Here are the current links with the right spacing:
<img width="192" alt="Screenshot 2023-08-29 at 5 53 05 PM" src="https://github.com/Strategy11/business-directory-plugin/assets/1116876/4edd21a6-e918-4907-b149-af7d36186e90">

But this is spaced too high. It should be the same 16px as the above screenshot:
<img width="355" alt="Screenshot 2023-08-29 at 5 53 52 PM" src="https://github.com/Strategy11/business-directory-plugin/assets/1116876/4a1821dc-684b-4e85-9767-b86f3e2f9be3">

stephywells reviewed (changes requested):
This is working great! Nice job @shervElmi !

I'm seeing a few small styling issues, but it's super close.

shervElmi commented:
Thank you for your review @stephywells. I hadn't realized you were on this PR; my apologies for that oversight!

shervElmi commented:
Thank you for your input on this, @stephywells and @mostafa-hisham. I've addressed the issues you mentioned. Please let me know if you need any further changes.

stephywells reviewed (changes requested):
Thanks @shervElmi this is almost there now. I'm just checking the last details.
<img width="566" alt="Screenshot 2023-08-30 at 2 01 15 PM" src="https://github.com/Strategy11/business-directory-plugin/assets/1116876/766904e2-5ec1-4036-9069-87505485c885">

Can we update the colors to match BD too?
--floating-links-bg-hover-color: rgba(109,135,185,0.07);
--floating-links-hover-color: #3C4B5D;
.s11-floating-links-nav-text{color: rgba(60,75,93,0.8);}

Can we also set the X to rgba(60,75,93,0.8)? We could set fill="currentColor" in the svg so we can change it in the css.

Percie feels a little too small to me. This is a mistake in the design with extra white space in the svg. Can we change the svg from:
width="31" height="40"
to:
width="35" height="46"

Also, I'm not seeing any utm on the links. Can we add those so we know if this is helping or not? 
- In FF, the links to our site should run through FrmAppHelper::admin_upgrade_link().
- In BD, it should go through wpbdp_admin_upgrade_link().

With these fixes, it should be completely ready to go. Thanks for fixing these @shervElmi !

stephywells commented on includes/admin/helpers/class-notices.php:
```diff
@@ -30,11 +30,6 @@ public static function show_bell() {
 			<ul class="wpbdp-bell-notifications-list"></ul>
 		</div>
-		<div class="wpbdp-bell-notification">
-			<a class="wpbdp-bell-notification-icon" href="#">
```
I'm seeing some leftover references to ".wpbdp-bell-notification" and "wpbdp-bell-notification-icon". Can we remove more of this?

stephywells replied:
I'm fine with that. Thank you!

stephywells replied:
@shervElmi I'm seeing css that isn't being used that would be nice to remove. It's much easier to do now than later.

stephywells reviewed (changes requested):
Well... I was wrong. I have a few more requests. I hadn't reviewed the code in detail. Sorry about that!

shervElmi commented on includes/class-assets.php:
```diff
@@ -485,4 +488,35 @@ private static function is_jquery_ui_css_loaded() {
+
+		// Enqueue the Floating Links scripts.
+		wp_enqueue_script( 's11-floating-links-notifications', $plugin_url . '/js/packages/floating-links/s11-floating-links-notifications.js', array(), $version, true );
+		wp_enqueue_script( 's11-floating-links', $plugin_url . '/js/packages/floating-links/s11-floating-links.js', array(), $version, true );
```
Of course, I'll incorporate your input. My goal is to make updates easier to manage across multiple plugins. Since we don't have notifications in FF, I've decided to create a separate file that we can easily copy and paste if needed. The idea is to only change the `config.js` file and keep `s11-floating-links.js` and `s11-floating-links-notifications.js` the same across all plugins, maintaining consistent functionality.

shervElmi replied:
We can ignore this for now because I've researched how we could combine them. Implementing this would require some setup in our plugin, which could be time-consuming. Below are the structures for your consideration. Please let me know what your thoughts are. Do you think we should proceed with this?

1. **Run the following commands to install `grunt-babel`, `@babel/core`, and `@babel/preset-env`:**

    ```bash
    npm install grunt-babel @babel/core @babel/preset-env --save-dev
    ```

2. **Update your `Gruntfile.js` to include a new Babel task. You'll add the following section:**

    ```javascript
    // Add these lines at the top of your Gruntfile.js
    const babel = require('grunt-babel');
    
    // ...existing code
    
    module.exports = function( grunt ) {
        // ...existing config
    
        // Add this inside the 'grunt.initConfig' function
        grunt.config.set('babel', {
            options: {
                sourceMap: true,
                presets: ['@babel/preset-env']
            },
            dist: {
                files: [
                    {
                        expand: true,
                        cwd: 'src/js/',  // replace with your source JavaScript files path
                        src: ['*.js'],
                        dest: 'dist/js/',  // replace with your transpiled files destination folder
                        ext: '.js'
                    }
                ]
            }
        });
    
        // ...existing tasks
    
        // Load grunt-babel
        grunt.loadNpmTasks('grunt-babel');
    
        // ...existing code
    };
    ```

3. **Add `babel` to your task lists wherever you'd like it to run. For example:**

    ```javascript
    // Inside your 'minify' task or any other task where you want to run Babel
    if ('undefined' !== typeof grunt.config.get('babel')) {
        grunt.task.run('babel');
    }
    ```

stephywells replied:
@shervElmi I don't think this would take much time, but I don't think we need to make it a priority right now. We can skip it since this is only on BD pages in the admin.

shervElmi commented on includes/class-assets.php:
```diff
@@ -515,7 +511,17 @@ private static function enqueue_floating_links( $plugin_url, $version ) {
+					'documentation' => wpbdp_admin_upgrade_link( 'floating-links', 'get-help/' ),
+				),
+				'proVersion'  => array(
+					'support_and_docs' => wpbdp_admin_upgrade_link( 'floating-links', 'get-help/' ),
```
I've pushed a commit addressing your feedback @stephywells and covering all items, except for a few where I've added comments.
I have another question: When a user activates the premium version, they will be sent a link containing the `utm_campaign=liteplugin` parameter for tracking. Does this look good to you, or should it be changed?

stephywells replied:
That's fine, since it's still in the lite plugin code. Thanks for checking!

stephywells commented on assets/js/packages/floating-links/config.js:
```diff
@@ -0,0 +1,158 @@
+		hoverColor: '#3C4B5D',
+		bgHoverColor: 'rgba(109,135,185,0.07)',
+		logoIcon: `
+		<svg width="35" height="46" viewBox="0 0 31 40" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Looks like we can cut down the svg size quite a bit with https://jakearchibald.github.io/svgomg/

stephywells reviewed (approved):
This matches great, works great, and looks great!

Very nice work on this @shervElmi, and thinking through all the complexities of keeping it easy to maintain for all the products. Thanks so much!

I think we can merge this once my last comment is addressed on the SVG compression.

shervElmi commented:
@stephywells, I appreciate your positive words. Thank you!
I just used the mentioned site to reduce the logo icon size.

stephywells commented:
Thank you @shervElmi ! It looks great!

stephywells commented:
@mostafa-hisham This one is just waiting on your review now.

stephywells commented:
Yay! Go team!
